drivers/aks/aks_driver.go

```diff
@@ -562,32 +920,141 @@ func (d *Driver) createOrUpdate(ctx context.Context, options *types.DriverOption
+		"westindia":          "centralindia",
+		"koreacentral":       "southeastasia",
+		"koreasouth":         "southeastasia",
+		"francecentral":      "westeurope",
+		"francesouth":        "westeurope",
+	}
+
+	workspaceRegion := regionToOmsRegionMap[state.Location]
+	workspaceRegionCode := locationToOmsRegionCodeMap[workspaceRegion]
+
+	workspaceResourceGroup := state.LogAnalyticsWorkspaceResourceGroup
+	if workspaceResourceGroup == "" {
+		workspaceResourceGroup = state.ResourceGroup
+	}
+
+	workspaceName := state.LogAnalyticsWorkspace
+	if workspaceName == "" {
+		workspaceName = fmt.Sprintf("%s-%s-%s", state.ResourceGroup, state.SubscriptionID, workspaceRegionCode)
+	}
+
+	if gotRet, gotErr := client.Get(ctx, workspaceResourceGroup, workspaceName); gotErr == nil {
+		return *gotRet.ID, nil
+	}
+
+	logrus.Info("Create Azure Log Analytics Workspace %q on Resource Group %q", workspaceName, workspaceResourceGroup)
+
+	asyncRet, asyncErr := client.CreateOrUpdate(ctx, workspaceResourceGroup, workspaceName, operationalinsights_20151101preview.Workspace{
+		Location: to.StringPtr(workspaceRegion),
+		WorkspaceProperties: &operationalinsights_20151101preview.WorkspaceProperties{
+			Sku: &operationalinsights_20151101preview.Sku{
+				Name: operationalinsights_20151101preview.Standalone,
+			},
+		},
+	})
+	if asyncErr != nil {
+		return "", asyncErr
+	}
+
+	err = wait.Poll(5*time.Second, 30*time.Second, func() (bool, error) {
+		ret, err := asyncRet.Result(*client)
+		if err != nil {
+			return false, err
+		}
+
+		workspaceID = *ret.ID
+		return true, nil
+	})
+
+	return
+}
+
+func (d *Driver) resourceGroupExists(ctx context.Context, client *resources_20170510.GroupsClient, groupName string) (bool, error) {
 	resp, err := client.CheckExistence(ctx, groupName)
 	if err != nil {
-		return false, fmt.Errorf("error getting resource group %v: %v", groupName, err)
+		return false, fmt.Errorf("error getting Resource Group '%s': %v", groupName, err)
 	}
 
 	return resp.StatusCode == 204, nil
 }
 
-func (d *Driver) createResourceGroup(ctx context.Context, client *resources.GroupsClient, state state) error {
-	_, err := client.CreateOrUpdate(ctx, state.ResourceGroup, resources.Group{
-		Name:     to.StringPtr(state.ResourceGroup),
-		Location: to.StringPtr(state.Location),
+func (d *Driver) createResourceGroup(ctx context.Context, client *resources_20170510.GroupsClient, state *state) error {
+	resourceGroup, location := state.ResourceGroup, state.Location
+
+	_, err := client.CreateOrUpdate(ctx, resourceGroup, resources_20170510.Group{
+		Name:     to.StringPtr(resourceGroup),
+		Location: to.StringPtr(location),
 	})
 	if err != nil {
-		return fmt.Errorf("error creating resource group %v: %v", state.ResourceGroup, err)
+		return fmt.Errorf("error creating Resource Group '%s': %v", resourceGroup, err)
 	}
 
 	return nil
 }
 
-func storeState(info *types.ClusterInfo, state state) error {
+func storeState(info *types.ClusterInfo, state *state) error {
```

Comment:
Why did we change state to a pointer?

Reply:
I think we don't wan to copy the `state` again in ram.

Reply:
I would in general be against such a change unless we had a clear performance goal.  It seems like X kilobytes of savings is not worth a potential bug from changing the behavior of state.

Reply:
passing the pointer can be more expensive @thxCode. It causes the program to have to allocate to the heap instead of the stack. Generally prefer passing the struct.

Reply:
Rollback